frame/staking/src/voter_list.rs

```diff
@@ -0,0 +1,295 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provide a linked list of nominators, sorted by stake.
+
+use crate::{
+	slashing::SlashingSpans, AccountIdOf, Config, Nominations, Nominators, Pallet, VotingDataOf,
+	VoteWeight,
+};
+use codec::{Encode, Decode};
+use frame_support::{DefaultNoBound, StorageMap, StorageValue};
+use sp_runtime::SaturatedConversion;
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+	#[error("attempted to insert into the wrong position in the list")]
+	WrongInsertPosition,
+}
+
+/// Type of voter.
+///
+/// Similar to [`crate::StakerStatus`], but somewhat more limited.
+#[derive(Clone, Copy, Encode, Decode, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum VoterType {
+	Validator,
+	Nominator,
+}
+
+/// Fundamental information about a voter.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Voter<AccountId> {
+	/// Account Id of this voter
+	pub id: AccountId,
+	/// Whether the voter is a validator or nominator
+	pub voter_type: VoterType,
+	/// The current slashable balance this voter brings.
+	///
+	/// This value is cached because the actual vote weight is likely to change frequently due to
+	/// various rewards and slashes, without substantially affecting the overall position in the
+	/// list. It's cheaper and easier to operate on a cache than to look up the actual value for
+	/// each voter each time.
+	pub cache_weight: VoteWeight,
+}
+
+pub type VoterOf<T> = Voter<AccountIdOf<T>>;
+
+impl<T: Config> Pallet<T> {
+	/// `Self` accessor for `NominatorList<T>`
+	pub fn voter_list() -> VoterList<T> {
+		VoterList::<T>::get()
+	}
+}
+
+/// Linked list of nominstors, sorted by stake.
+#[derive(DefaultNoBound, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct VoterList<T: Config> {
+	head: Option<AccountIdOf<T>>,
+	tail: Option<AccountIdOf<T>>,
+}
```

Comment:
```suggestion
pub struct DoubleLinkedList<T: ...> {
	head: Option<T>,
	tail: Option<T>,
}
```

don't want to make things too hard for you, but this might help in the future. Then you can just call it something generic like `DoubleLinkedList`.

Reply:
but I am also fine with doing this only tailored for staking now, and later extract it.

Reply:
I'm inclined to keep things specific for now and extract it later. Given that we need multiple storage items, I don't yet have a good sense of how we'd generalize this.